unified-scheduler-pool/src/lib.rs

```diff
@@ -0,0 +1,749 @@
+        log_messages_bytes_limit: Option<usize>,
+        transaction_status_sender: Option<TransactionStatusSender>,
+        replay_vote_sender: Option<ReplayVoteSender>,
+        prioritization_fee_cache: Arc<PrioritizationFeeCache>,
+    ) -> Arc<Self> {
+        Arc::new_cyclic(|weak_self| Self {
+            schedulers: Mutex::default(),
+            log_messages_bytes_limit,
+            transaction_status_sender,
+            replay_vote_sender,
+            prioritization_fee_cache,
+            weak_self: weak_self.clone(),
+            next_scheduler_id: AtomicSchedulerId::new(PRIMARY_SCHEDULER_ID),
+            _phantom: PhantomData,
+        })
+    }
+
+    pub fn new_dyn(
+        log_messages_bytes_limit: Option<usize>,
+        transaction_status_sender: Option<TransactionStatusSender>,
+        replay_vote_sender: Option<ReplayVoteSender>,
+        prioritization_fee_cache: Arc<PrioritizationFeeCache>,
+    ) -> InstalledSchedulerPoolArc {
+        Self::new(
+            log_messages_bytes_limit,
+            transaction_status_sender,
+            replay_vote_sender,
+            prioritization_fee_cache,
+        )
+    }
+
+    // See a comment at the weak_self field for justification of this.
+    fn self_arc(&self) -> Arc<Self> {
+        self.weak_self
+            .upgrade()
+            .expect("self-referencing Arc-ed pool")
+    }
+
+    fn new_scheduler_id(&self) -> SchedulerId {
+        self.next_scheduler_id.fetch_add(1, Relaxed)
+    }
+
+    fn return_scheduler(&self, scheduler: Box<S>) {
+        assert!(!scheduler.has_context());
+
+        self.schedulers
+            .lock()
+            .expect("not poisoned")
+            .push(scheduler);
+    }
+
+    fn do_take_scheduler(&self, context: SchedulingContext) -> Box<S> {
+        // pop is intentional for filo, expecting relatively warmed-up scheduler due to having been
+        // returned recently
+        if let Some(mut scheduler) = self.schedulers.lock().expect("not poisoned").pop() {
+            scheduler.replace_context(context);
+            scheduler
+        } else {
+            Box::new(S::spawn(
+                self.self_arc(),
+                context,
+                TH::recreate_handler_with_pool(self),
+            ))
+        }
+    }
+}
+
+impl<S: SpawnableScheduler<TH>, TH: TaskHandler> InstalledSchedulerPool for SchedulerPool<S, TH> {
+    fn take_scheduler(&self, context: SchedulingContext) -> Box<dyn InstalledScheduler> {
+        self.do_take_scheduler(context)
+    }
+}
+
+pub trait TaskHandler: Send + Sync + Debug + Sized + 'static {
+    fn recreate_handler_with_pool<S: SpawnableScheduler<Self>>(
+        pool: &SchedulerPool<S, Self>,
+    ) -> Self;
+
+    fn handle<S: SpawnableScheduler<Self>>(
+        &self,
+        result: &mut Result<()>,
+        timings: &mut ExecuteTimings,
+        bank: &Arc<Bank>,
+        transaction: &SanitizedTransaction,
+        index: usize,
+        pool: &SchedulerPool<S, Self>,
```

Comment:
similarly, here this seems odd to me to pass the pool this far down the chain.
It appears the pool is used here to paass some shared context for execution - channels for status/vote, log mesage limit, fee cache, etc.
Is there anything on the pool **specifically** we require in here? If not, I'd reccomend splitting out some sort of `ExecutionContext` (possibly owned by pool?) which is passed here instead of the pool itself.

That would probably simplify things a great deal, as there's no cyclic dependency in the traits.

Reply:
> Is there anything on the pool specifically we require in here?

nice question! the answer is no: 584bd35aae420879bc5265c86299a4f1eddf9fcf

> That would probably simplify things a great deal, as there's no cyclic dependency in the traits.

note that the scattered `TH` things didn't go with the commit... to be object-safe, we still need to parametrize the struct itself not just the `schedule_execution` method... nevertheless, this commit simplifies the code and should alleviate human confusion... ;)